crates/kitsune_p2p/proxy/src/tx2.rs

```diff
@@ -0,0 +1,598 @@
+#![allow(clippy::new_ret_no_self)]
+#![allow(clippy::blocks_in_if_conditions)]
+//! Next-gen performance kitsune transport proxy
+
+use crate::*;
+use futures::future::BoxFuture;
+use futures::stream::{Stream, StreamExt};
+use kitsune_p2p_types::tx2::tx2_backend::*;
+use kitsune_p2p_types::tx2::tx2_frontend::tx2_frontend_traits::*;
+use kitsune_p2p_types::tx2::tx2_frontend::*;
+use kitsune_p2p_types::tx2::tx2_utils::*;
+use kitsune_p2p_types::tx2::*;
+use kitsune_p2p_types::*;
+use std::collections::HashMap;
+
+/// Wrap a tx2 backend transport with proxy logic.
+pub fn tx2_proxy(sub_fact: EpFactory, tls_config: TlsConfig) -> EpFactory {
+    ProxyEpFactory::new(sub_fact, tls_config)
+}
+
+// -- private -- //
+
+const PROXY_HELLO_DIGEST: u8 = 0x20;
+const PROXY_FWD_MSG: u8 = 0x30;
+
+struct ProxyConHnd {
+    uniq: Uniq,
+    sub_con: ConHnd,
+    local_digest: CertDigest,
+    remote_digest: CertDigest,
+}
+
+impl std::fmt::Debug for ProxyConHnd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ConHnd").finish()
+    }
+}
+
+impl ProxyConHnd {
+    pub fn new(sub_con: ConHnd, local_digest: CertDigest, remote_digest: CertDigest) -> ConHnd {
+        let uniq = Uniq::default();
+        let con = Self {
+            uniq,
+            sub_con,
+            local_digest,
+            remote_digest,
+        };
+        ConHnd(Arc::new(con))
+    }
+}
+
+fn promote_addr(base_addr: &TxUrl, cert_digest: &CertDigest) -> KitsuneResult<TxUrl> {
+    Ok(ProxyUrl::new(base_addr.as_str(), cert_digest.clone())
+        .map_err(KitsuneError::other)?
+        .as_str()
+        .into())
+}
+
+impl AsConHnd for ProxyConHnd {
+    fn uniq(&self) -> Uniq {
+        self.uniq
+    }
+
+    fn is_closed(&self) -> bool {
+        self.sub_con.is_closed()
+    }
+
+    fn close(&self, _code: u32, _reason: &str) -> BoxFuture<'static, ()> {
+        // TODO - FIXME
+        // we don't want to close the underlying sub_con,
+        // it could be shared for proxying...
+        // do we want to do *anything*?
```

Comment:
How do we close down the proxy when we are actually finished though?

Reply:
The close at the endpoint level will close everything properly... there just isn't yet a way to close down individual connections.

We probably want to fix this somehow at some point, but I don't think it's dire. Will leave the `TODO - FIXME` in there.